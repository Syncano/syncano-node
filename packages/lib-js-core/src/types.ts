export interface SyncanoResponse<T> {
  next: string | null
  prev: string | null
  objects: T[]
}

export interface ACL {
  [className: string]: string[] | {
    [id: string]: string[]
  }
}

export interface User {
  id: number
  username: string
  user_key: string
  created_at: string
  updated_at: string
}

export interface UserGroup {
  id: number
  name: string
  label: string
  description: string
}

export interface ClassObject {
  id: number
  created_at: string
  updated_at: string
  revision: number
}

export interface ClassObjectLinks {
  self: string
  [entity: string]: string
}

export interface AccountOwner {
  id: number
  email: string
  first_name: string
  last_name: string
  is_active: boolean
  has_password: boolean
  metadata: object
}

export interface LoginData extends AccountOwner {
  account_key: string
}

export interface BackupObject {
  id: number
  instance: string
  created_at: string
  updated_at: string
  size: number | null
  status: string
  status_info: string
  description: string
  label: string
  author: AccountOwner
  details: BackupDetails
  metadata: any
  links: ClassObjectLinks
}

export interface BackupDetails {
  response_template: {
    count: number
    list?: Array<string|number>
  }
  class: {
    count: number
    list?: Array<string|number>
  }
  socket: {
    count: number
    list?: Array<string|number>
  }
  script: {
    count: number
    list?: Array<string|number>
  }
  data_object: {
    count: number
    list?: Array<string|number>
  }
  hosting: {
    count: number
    list?: Array<string|number>
  }
  hosting_file: {
    count: number
    list?: Array<string|number>
  }
}

export interface Trace {
  meta: {
    REQUEST_METHOD: string
    PATH_INFO: string
    HTTP_HOST: string
    HTTP_CONNECTION: string
    HTTP_UPGRADE_INSECURE_REQUESTS: string
    HTTP_USER_AGENT: string
    HTTP_ACCEPT: string
    HTTP_ACCEPT_ENCODING: string
    HTTP_ACCEPT_LANGUAGE: string
    HTTP_COOKIE: string
    REMOTE_ADDR: string
  }
  id: number
  status: string
  executed_at: string
  duration: number
  links: {
    self: string
    [x: string]: string
  }
}

export interface Socket {
  name: string
  description: string
  created_at: string
  updated_at: string
  version: string
  status: string
  status_info: any
  install_url: any
  metadata: {
    name: string
    keywords: string[]
    sources: {
      [x: string]: string
    }
  }
  config: object
  installed: {
    endpoints: {
      [endpointName: string]: {
        script: string
        runtime: string
      }
    }
    classes: {
      [className: string]: {
        [fieldName: string]: string
      }
    }
  }
  files: {
    [fileName: string]: {
        checksum: string
        size: number
        file: string
        helper: boolean
    }
  }
  environment: string
  links: {
    self: string
    update: string
    endpoints: string
    handlers: string
    zip_file: string
    [x: string]: string
  }
}

export interface Settings {
  meta: any
  token?: string
  socket?: string
  instanceName?: string
  setResponse?: any
  HttpResponse?: any
}

export interface InstanceOwner extends AccountOwner {
  metadata: InstanceMetadata
}

export interface InstanceMetadata {
  icon?: string
  color?: string
  [x: string]: any
}

export interface Instance {
  name: string
  description: string
  owner: InstanceOwner
  created_at: string
  updated_at: string
  role: string
  location: string
  metadata: InstanceMetadata
  links: {
    self: string
    admins: string
    snippets: string
    endpoints: string
    push_notification: string
    classes: string
    invitations: string
    api_keys: string
    triggers: string
    schedules: string
    users: string
    groups: string
    channels: string
    batch: string
    rename: string
    backups: string
    restores: string
    hosting: string
    'classes-acl': string
    'channels-acl': string
    'script-endpoints-acl': string
    'groups-acl': string
    'users-schema': string
    'triggers-emit': string
    sockets: string
    'sockets-install': string
    environments: string
    [x: string]: string
  }
}

export interface Hosting {
  name: string
  is_default: boolean
  description: string
  created_at: string
  updated_at: string
  domains: string[]
  is_active: boolean
  ssl_status: string
  auth: object
  config: {
    browser_router: boolean
  }
  links: {
    self: string
    files: string
    set_default: string
    enable_ssl: string
    [x: string]: string
  }
}

export interface HostingFile {
  id: number
  path: string
  size: number
  checksum: string
  links: {
    self: string
    [x: string]: string
  }
}

export interface SyncanoClass {
  name: string
  description: string
  schema: SchemaObject[]
  status: string
  created_at: string
  updated_at: string
  objects_count: number
  revision: number
  metadata: object
}

export interface SchemaObject {
  name: string
  type: 'reference'|'relation'|'text'|'string'|'file'|'object'|'array'|'geopoint'|'integer'|'float'|'boolean'|'datetime'
  order_index?: boolean
  filter_index?: boolean
  unique?: boolean
  target?: string
}

export interface ChannelResponse<T> {
  id: number
  room: string
  created_at: string
  action: string
  author: {
    admin: number
  }
  metadata: {
    type: string
  }
  payload: T
  links: {
    self: string
    [x: string]: string
  }
}
