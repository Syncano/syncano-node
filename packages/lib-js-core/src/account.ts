import { ResponseError } from './errors'
import QueryBuilder from './query-builder'

export interface AccountData {
  id: number
  email: string
  first_name: string
  last_name: string
  is_active: boolean
  has_password: boolean
  metadata: object
}

export interface LoginData extends AccountData {
  account_key: string
}

/**
 * Syncano account query builder
 * @property {Function}
 */
export class Account extends QueryBuilder {
  /**
   * Get details of Syncano account
   */
  public get (authKey: string): Promise<AccountData> {
    const fetch = this.nonInstanceFetch.bind(this)
    return new Promise((resolve, reject) => {
      const headers = {
        'X-API-KEY': authKey
      }

      fetch(this.url(), {}, headers)
        .then((res: AccountData)  => resolve(res))
        .catch((err: ResponseError) => reject(err))
    })
  }

  public login (
    {email, password}: {email: string, password: string}
  ): Promise<LoginData> {
    const fetch = this.nonInstanceFetch.bind(this)
    return new Promise((resolve, reject) => {
      const options = {
        body: JSON.stringify({ email, password }),
        method: 'POST'
      }
      fetch(`${this.url()}auth/`, options)
        .then((res: LoginData) => resolve(res))
        .catch((err: ResponseError) => reject(err))
    })
  }

  public register (
    {email, password}: {email: string, password: string}
  ): Promise<LoginData> {
    const fetch = this.nonInstanceFetch.bind(this)
    return new Promise((resolve, reject) => {
      const options = {
        body: JSON.stringify({ email, password }),
        method: 'POST'
      }

      fetch(`${this.url()}register/`, options)
        .then((res: LoginData) => resolve(res))
        .catch((err: ResponseError) => reject(err))
    })
  }

  private url () {
    return `${this.getSyncanoURL()}/account/`
  }
}

export default Account
