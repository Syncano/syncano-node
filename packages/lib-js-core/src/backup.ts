import { NotFoundError } from './errors'
import QueryBuilder from './query-builder'
import { BackupObject } from './types'

/**
 * Syncano Backups
 *
 * @property {Function}
 */
class Backup extends QueryBuilder {
  /**
   * Backups list
   */
  public async list (): Promise<BackupObject[]> {
    return this.getBackups(this.url())
  }

  /**
   * Create new backup
   */
  public async create (): Promise<BackupObject> {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'POST'
    }
    const headers = {
      'X-API-KEY': this.instance.accountKey
    }

    return fetch(this.url(), options, headers)
  }

  /**
   * Delete backup
   */
  public async delete (id?: number | number[]): Promise<any> {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'DELETE'
    }
    const headers = {
      'X-API-KEY': this.instance.accountKey
    }

    if (id === undefined) {
      const items = await this.list()
      const ids = items.map((item: BackupObject) => item.id)

      return this.delete(ids)
    }

    if (Array.isArray(id)) {
      return Promise.all(
        id.map((item) =>
          fetch(`${this.url()}${item}/`, options, headers).then(() => true).catch(() => false)
        )
      )
    }

    await fetch(`${this.url()}${id}/`, options, headers)

    return undefined
  }

  /**
   * Get backup
   *
   * @deprecated User `find`
   */
  public async get (id: number): Promise<BackupObject | undefined> {
    return this.find(id)
  }

  /**
   *  Get backup by id
   *
   * @example
   * const item = await backup.find(10)
   */
  public async find(id: number): Promise<BackupObject | undefined> {
    try {
      return this.findOrFail(id)
    } catch (err) {
      return undefined
    }
  }

  /**
   *  Get backup by id or throw error
   *
   * @example
   * const item = await backup.findOrFail(10)
   */
  public async findOrFail(id: number): Promise<BackupObject> {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'GET'
    }
    const headers = {
      'X-API-KEY': this.instance.accountKey
    }

    try {
      return fetch(`${this.url()}${id}/`, options, headers)
    } catch (err) {
      throw new NotFoundError(`Backup with id: "${id}" was not found.`)
    }
  }

  /**
   *  Get list of backups by id
   *
   * @example
   * const items = await backup.findMany([10, 11])
   */
  public async findMany(ids: number[]): Promise<BackupObject[]> {
    const groups = await this.list()

    return groups.filter((item) => ids.includes(item.id))
  }

  /**
   * Get last backup
   */
  public async last () {
    const backups = await this.list()

    return backups.pop()
  }

  protected url () {
    const {instanceName} = this.instance

    return `${this.getInstanceURL(instanceName)}/backups/full/`
  }

  protected async getBackups (url: string) {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'GET'
    }
    const headers = {
      'X-API-KEY': this.instance.accountKey
    }
    const res = await fetch(url, options, headers)
    let backups = res.objects

    if (res.next) {
      const nextBackups = await this.getBackups(`${this.baseUrl}${res.next}`)

      backups = backups.concat(nextBackups.objects)
    }

    return backups
  }
}

export default Backup
