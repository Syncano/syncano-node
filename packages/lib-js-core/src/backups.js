import QueryBuilder from './query-builder'

/**
 * Syncano Backups
 *
 * @property {Function}
 */
class Backups extends QueryBuilder {
  url () {
    const {instanceName} = this.instance

    return `${this._getInstanceURL(instanceName)}/backups/full/`
  }

  /**
   * Backups list helper
   *
   * @param {string} url
   *
   * @return {Array}
   */
  async _getBackups (url) {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'GET',
    }
    const headers = {
      'x-api-key': this.instance.accountKey
    }
    const res = await fetch(url, options, headers)
    let backups = res.objects

    if(res.next) {
      backups = backups.concat(await this._getGroups(`${this.baseUrl}${res.next}`))
    }

    return backups
  }

  /**
   * Backups list
   *
   * @return {Promise}
   */
  async list () {
    return  this._getBackups(this.url())
  }

  /**
   * Create new backup
   *
   * @return {Promise}
   */
  async create () {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'POST',
    }
    const headers = {
      'x-api-key': this.instance.accountKey
    }

    return fetch(this.url(), options, headers)
  }

  /**
   * Delete backup
   *
   * @param {number} id - Backup id
   *
   * @return {Promise}
   */
  async delete (id) {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'DELETE',
    }
    const headers = {
      'x-api-key': this.instance.accountKey
    }

    return fetch(`${this.url()}${id}/`, options, headers)
  }

  /**
   * Delete all backups
   *
   * @return {Promise}
   */
  async deleteAll () {
    const backups = await this.list()
    backups.forEach(({id}) => this.delete(id))
  }

  /**
   * Get backup
   *
   * @param {number} id - Backup id
   *
   * @return {Promise}
   */
  async get (id) {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'GET',
    }
    const headers = {
      'x-api-key': this.instance.accountKey
    }

    return fetch(`${this.url()}${id}/`, options, headers)
  }

  /**
   * Get last backup
   *
   * @return {Promise}
   */
  async last () {
    const backups = await this.list()
    return backups.pop()
  }
}

export default Backups
