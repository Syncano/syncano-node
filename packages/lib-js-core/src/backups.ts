import QueryBuilder from './query-builder'
import { BackupObject } from './types'

/**
 * Syncano Backups
 *
 * @property {Function}
 */
class Backups extends QueryBuilder {
  /**
   * Backups list
   */
  public async list (): Promise<BackupObject[]> {
    return this.getBackups(this.url())
  }

  /**
   * Create new backup
   */
  public async create (): Promise<BackupObject> {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'POST'
    }
    const headers = {
      'X-API-KEY': this.instance.accountKey
    }

    return fetch(this.url(), options, headers)
  }

  /**
   * Delete backup
   */
  public async delete (id?: number | number[]): Promise<any> {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'DELETE'
    }
    const headers = {
      'X-API-KEY': this.instance.accountKey
    }

    if (id === undefined) {
      const items = await this.list()
      const ids = items.map((item: BackupObject) => item.id)

      return this.delete(ids)
    }

    if (Array.isArray(id)) {
      return Promise.all(
        id.map((item) =>
          fetch(`${this.url()}${item}/`, options, headers)
        )
      )
    }

    return fetch(`${this.url()}${id}/`, options, headers)
  }

  /**
   * Delete all backups
   *
   * @deprecated Use `delete` method without arguments
   */
  public async deleteAll () {
    return this.delete()
  }

  /**
   * Get backup
   */
  public async get (id: number): Promise<BackupObject> {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'GET'
    }
    const headers = {
      'X-API-KEY': this.instance.accountKey
    }

    return fetch(`${this.url()}${id}/`, options, headers)
  }

  /**
   * Get last backup
   */
  public async last () {
    const backups = await this.list()

    return backups.pop()
  }

  protected url () {
    const {instanceName} = this.instance

    return `${this.getInstanceURL(instanceName)}/backups/full/`
  }

  /**
   * Backups list helper
   */
  protected async getBackups (url: string) {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'GET'
    }
    const headers = {
      'X-API-KEY': this.instance.accountKey
    }
    const res = await fetch(url, options, headers)
    let backups = res.objects

    if (res.next) {
      const nextBackups = await this.getBackups(`${this.baseUrl}${res.next}`)

      backups = backups.concat(nextBackups.objects)
    }

    return backups
  }
}

export default Backups
