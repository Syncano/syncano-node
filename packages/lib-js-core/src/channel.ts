import QueryBuilder from './query-builder'

export interface ChannelResponse<T> {
  id: number,
  room: string,
  created_at: string,
  action: string,
  author: {
    admin: number
  },
  metadata: {
    type: string
  },
  payload: T,
  links: {
    self: string
    [x: string]: string
  }
}

export class Channel extends QueryBuilder {
  /**
   * Publish any payload to channel
   *
   * @param channelName A string value used as channel name. Channel types:
   *                    - Public channels: `messages`
   *                    - User private channel: `messages.${username}`
   *                    - Channel room: `messages.${room}`
   * @param payload Value published to channel
   */
  public publish<T> (channelName: string, payload: T): Promise<ChannelResponse<T>> {
    const fetch = this.fetch.bind(this)

    return new Promise((resolve, reject) => {
      const options = {
        body: JSON.stringify({room: channelName, payload}),
        method: 'POST'
      }
      fetch(this.url(), options)
        .then(resolve)
        .catch(reject)
    })
  }

  private url () {
    const {instanceName} = this.instance

    return `${this.getInstanceURL(instanceName)}/channels/default/publish/`
  }
}

export default Channel
