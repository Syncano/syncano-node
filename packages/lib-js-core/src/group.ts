import {NotFoundError} from './errors'
import QueryBuilder from './query-builder'
import {SyncanoResponse, UserGroup} from './types'

export type Primitive = undefined | null | boolean | string | number

/**
 * Syncano groups
 * @property {Function}
 */
class Group extends QueryBuilder {
  /**
   * Get groups list
   *
   * @example
   * const groupsList = await groups.list()
   */
  public async list(): Promise<UserGroup[]> {
    return this.getGroups(this.url())
  }

  /**
   *  Get group by label
   *
   * @example
   * const group = await groups.find(10)
   * const group = await groups.find('admin')
   */
  public async find(groupIdOrName: string | number): Promise<UserGroup | undefined> {
    try {
      return this.findOrFail(groupIdOrName)
    } catch (err) {
      return undefined
    }
  }

  /**
   *  Get group by label
   *
   * @example
   * const group = await groups.find(10)
   * const group = await groups.find('admin')
   */
  public async findOrFail(groupIdOrName: string | number): Promise<UserGroup> {
    const groups = await this.list()
    const paramName = typeof groupIdOrName === 'string' ? 'name' : 'id'
    let result

    if (typeof groupIdOrName === 'string') {
      result = groups.find((item) => item.name === groupIdOrName)
    }

    if (typeof groupIdOrName === 'number') {
      result = groups.find((item) => item.id === groupIdOrName)
    }

    if (!result) {
      throw new NotFoundError(`Group with ${paramName}: "${groupIdOrName}" was not found.`)
    }

    return result
  }

  /**
   *  Get group by label
   *
   * @example
   * const group = await groups.findMany([10, 11])
   * const group = await groups.findMany(['admin', 'manager'])
   */
  public async findMany(id: string[] | number[]): Promise<UserGroup[]> {
    const groups = await this.list()

    if ((id as any[]).every((item) => typeof item === 'string')) {
      return groups.filter(
        (item) => (id as string[]).includes(item.name)
      )
    }

    if ((id as any[]).every((item) => typeof item === 'number')) {
      return groups.filter(
        (item) => (id as number[]).includes(item.id)
      )
    }

    return []
  }

  /**
   * Create new group
   *
   * @example
   * const adminGroup = await groups.create({
   *   name: 'admin',
   *   label: 'Admin',
   *   description: 'Optional description'
   * })
   */
  public async create (group: {
    name: string
    label?: string
    description?: string
  }): Promise<UserGroup> {
    const fetch = this.fetch.bind(this)

    return fetch(this.url(), {
      method: 'POST',
      body: JSON.stringify(group)
    })
  }

  /**
   * update group
   *
   * @example
   * groups.update(2, {name: 'admin', label: 'Admin'})
   */
  public async update (id: number, data: {
    name?: string
    label?: string
    description?: string
  }): Promise<UserGroup> {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'PATCH',
      body: JSON.stringify(data)
    }

    return fetch(`${this.url()}${id}/`, options)
  }

  /**
   * Delete group
   *
   * @example
   * // Delete by id
   * groups.delete(2)
   * // Or delete by name
   * groups.delete('admin')
   */
  public async delete (groupIdOrName: number | string): Promise<undefined> {
    const fetch = this.fetch.bind(this)
    const options = {method: 'DELETE'}

    if (typeof groupIdOrName === 'string') {
      const group = await this.find(groupIdOrName)

      if (!group) {
        throw new NotFoundError(`Group with name: "${groupIdOrName}" was not found.`)
      }

      await fetch(`${this.url()}${group.id}/`, options)

      return
    }

    await fetch(`${this.url()}${groupIdOrName}/`, options)

    return
  }

  /**
   * Check if user belongs to a group
   *
   * @example
   * const isUserInGroup = await groups.isUserAttachedTo('admin', 10)
   */
  public async isUserAttachedTo(groupIdOrName: number | string, userId: number) {
    const groups = await this.getUserGroups(userId)
    const paramName = typeof groupIdOrName === 'string' ? 'name' : 'id'

    return groups.some(((item) => item[paramName] === groupIdOrName))
  }

  /**
   * Get list user groups
   *
   * @example
   * const userGroups = await groups.getUserGroups(1)
   */
  public async getUserGroups(userId: number) {
    const url = `${this.getInstanceURL(this.instance.instanceName)}/users/${userId}/groups/`
    const {objects}: SyncanoResponse<UserGroup> = await this.fetch.bind(this)(url)

    return objects
  }

  /**
   * Add user to group
   *
   * @example
   * // Attach by group id
   * groups.attachUser(1, 2)
   * // Attach by group name
   * groups.attachUser('admin', 1)
   */
  public async attachUser (groupIdOrName: number | string, userId: number) {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'POST',
      body: JSON.stringify({user: userId})
    }

    if (typeof groupIdOrName === 'string') {
      const group = await this.findOrFail(groupIdOrName)

      return fetch(`${this.url()}${group.id}/users/`, options)
    }

    return fetch(`${this.url()}${groupIdOrName}/users/`, options)
  }

  /**
   * Remove user from group
   *
   * @example
   * // Detach user by group id
   * groups.detachUser(2, 1)
   * // Detach user by group name
   * groups.detachUser('admin', 1)
   */
  public async detachUser (groupIdOrName: number | string, userId: number) {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'DELETE'
    }

    if (typeof groupIdOrName === 'string') {
      const group = await this.findOrFail(groupIdOrName)

      return fetch(`${this.url()}${group.id}/users/${userId}/`, options)
    }

    return fetch(`${this.url()}${groupIdOrName}/users/${userId}/`, options)
  }

  private async getGroups (url: string) {
    const fetch = this.fetch.bind(this)
    const options = {
      method: 'GET'
    }
    const res = await fetch(url, options)
    let groups = res.objects

    if (res.next) {
      groups = groups.concat(await this.getGroups(`${this.baseUrl}${res.next}`))
    }

    return groups
  }

  private url () {
    const {instanceName} = this.instance

    return `${this.getInstanceURL(instanceName)}/groups/`
  }
}

export default Group
