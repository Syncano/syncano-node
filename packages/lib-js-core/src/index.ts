import Server from './server'

export default Server
