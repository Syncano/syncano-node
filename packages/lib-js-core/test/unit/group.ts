import * as chai from 'chai'
import * as chaiAsPromised from 'chai-as-promised'
import * as nock from 'nock'
import Server from '../../src'
import {GroupClass} from '../../src/group'
import {SyncanoResponse, UserGroup} from '../../src/types'

chai.use(chaiAsPromised)
chai.should()

type ResponseTuple = [number, SyncanoResponse<UserGroup>]

describe('Group', () => {
  const instanceName = 'testInstance'
  let api: nock.Scope
  let groups: GroupClass

  const USER = {
    id: 2,
    username: 'testUser',
    revision: 6,
    user_key: 'xyz',
    created_at: '2018-07-16T16:01:15.736047Z',
    updated_at: '2018-07-31T15:16:31.259745Z'
  }
  const GROUPS_URL_V2 = `/v2/instances/${instanceName}/groups/`
  const USER_GROUPS_URL_V3 = `/v3/instances/${instanceName}/users/${USER.id}/groups/`
  const ADMIN_GROUP: UserGroup = {
    id: 1,
    name: 'admin',
    label: 'Admin',
    description: ''
  }
  const MANAGER_GROUP: UserGroup = {
    id: 2,
    name: 'manager',
    label: 'Manager',
    description: 'Manage stuff'
  }
  const LIST_RESPONSE: ResponseTuple = [200, {
    next: null,
    prev: null,
    objects: [ADMIN_GROUP, MANAGER_GROUP]
  }]

  beforeEach(() => {
    const server = new Server({
      token: 'testKey',
      instanceName
    })
    groups = server.groups
    api = nock(`https://${process.env.SYNCANO_HOST || 'api.syncano.io'}`)
  })

  describe('#find', () => {
    it('should find single group by id ', async () => {
      api.get(GROUPS_URL_V2).reply(...LIST_RESPONSE)

      const group = await groups.find(MANAGER_GROUP.id)

      expect(group).toMatchSnapshot()
    })

    it('should find single group by name', async () => {
      api.get(GROUPS_URL_V2).reply(...LIST_RESPONSE)

      const group = await groups.find(MANAGER_GROUP.name)

      expect(group).toMatchSnapshot()
    })
  })

  describe('#findMany', () => {
    it('should find many groups by id ', async () => {
      api
        .get(GROUPS_URL_V2).reply(...LIST_RESPONSE)
        .get(GROUPS_URL_V2).reply(...LIST_RESPONSE)

      const res = await groups.findMany([ADMIN_GROUP.id, MANAGER_GROUP.id])
      const res2 = await groups.findMany([MANAGER_GROUP.id])

      expect(res).toHaveLength(2)
      expect(res).toMatchSnapshot()
      expect(res2).toHaveLength(1)
      expect(res2).toMatchSnapshot()
    })

    it('should find many groups by name', async () => {
      api
        .get(GROUPS_URL_V2).reply(...LIST_RESPONSE)
        .get(GROUPS_URL_V2).reply(...LIST_RESPONSE)
      const group = await groups.find(MANAGER_GROUP.name)

      expect(group).toMatchSnapshot()
    })
  })

  describe('#list', () => {
    it('should get list of groups', async () => {
      api.get(GROUPS_URL_V2).reply(...LIST_RESPONSE)

      const list = await groups.list()

      expect(list).toMatchSnapshot()
    })
  })

  describe('#create', () => {
    it('should create a group', async () => {
      api.post(GROUPS_URL_V2).reply(200, ADMIN_GROUP)

      const group = await groups.create(ADMIN_GROUP)

      expect(group).toMatchSnapshot()
    })
  })

  describe('#update', () => {
    it('should update a group by id', () => {
      api.patch(`${GROUPS_URL_V2}${MANAGER_GROUP.id}/`).reply(200, ADMIN_GROUP)

      return expect(groups.update(MANAGER_GROUP.id, ADMIN_GROUP)).resolves.toMatchSnapshot()
    })
  })

  describe('#delete', () => {
    it('should delete a group by id', () => {
      api.delete(`${GROUPS_URL_V2}${ADMIN_GROUP.id}/`).reply(200)

      return expect(groups.delete(ADMIN_GROUP.id)).resolves.toMatchSnapshot()
    })
  })

  describe('#attachUser', () => {
    it('should attach user to a group with given id', () => {
      api.post(`${GROUPS_URL_V2}${ADMIN_GROUP.id}/users/`).reply(200, USER)

      return expect(groups.attachUser(ADMIN_GROUP.id, USER.id)).resolves.toMatchSnapshot()
    })

    it('should attach user to a group with given name', () => {
      api.get(GROUPS_URL_V2).reply(...LIST_RESPONSE)
      api.post(`${GROUPS_URL_V2}${ADMIN_GROUP.id}/users/`).reply(200, USER)

      return expect(groups.attachUser(ADMIN_GROUP.name, USER.id)).resolves.toMatchSnapshot()
    })
  })

  describe('#detachUser', () => {
    it('should detach user from a group with given id', () => {
      api.delete(`${GROUPS_URL_V2}${ADMIN_GROUP.id}/users/${USER.id}/`).reply(200)

      return expect(groups.detachUser(ADMIN_GROUP.id, USER.id)).resolves.toMatchSnapshot()
    })

    it('should detach user from a group with given name', () => {
      api.get(GROUPS_URL_V2).reply(...LIST_RESPONSE)
      api.delete(`${GROUPS_URL_V2}${ADMIN_GROUP.id}/users/${USER.id}/`).reply(200)

      return expect(groups.detachUser(ADMIN_GROUP.name, USER.id)).resolves.toMatchSnapshot()
    })
  })

  describe('#isUserAttahcedTo', () => {
    it('should return true when user belongs to a group', () => {
      api.get(USER_GROUPS_URL_V3).reply(...LIST_RESPONSE)

      return expect(groups.isUserAttachedTo(ADMIN_GROUP.id, USER.id)).resolves.toMatchSnapshot()
    })

    it('should return false when user doesn\'t belong to a group', () => {
      api.get(USER_GROUPS_URL_V3).reply(200, {
        next: null,
        prev: null,
        objects: [MANAGER_GROUP]
      })

      return expect(groups.isUserAttachedTo(ADMIN_GROUP.name, USER.id)).resolves.toMatchSnapshot()
    })
  })

  describe('#getUserGroups', () => {
    it('should return list of user groups', () => {
      api.get(USER_GROUPS_URL_V3).reply(...LIST_RESPONSE)

      return expect(groups.getUserGroups(USER.id)).resolves.toMatchSnapshot()
    })
  })
})
