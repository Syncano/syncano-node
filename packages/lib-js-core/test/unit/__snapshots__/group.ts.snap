// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Group #attachUser should attach user to a group with given id 1`] = `
Object {
  "created_at": "2018-07-16T16:01:15.736047Z",
  "id": 2,
  "revision": 6,
  "updated_at": "2018-07-31T15:16:31.259745Z",
  "user_key": "xyz",
  "username": "testUser",
}
`;

exports[`Group #attachUser should attach user to a group with given name 1`] = `
Object {
  "created_at": "2018-07-16T16:01:15.736047Z",
  "id": 2,
  "revision": 6,
  "updated_at": "2018-07-31T15:16:31.259745Z",
  "user_key": "xyz",
  "username": "testUser",
}
`;

exports[`Group #create should create a group 1`] = `
Object {
  "description": "",
  "id": 1,
  "label": "Admin",
  "name": "admin",
}
`;

exports[`Group #delete should delete a group by id 1`] = `undefined`;

exports[`Group #detachUser should detach user from a group with given id 1`] = `undefined`;

exports[`Group #detachUser should detach user from a group with given name 1`] = `undefined`;

exports[`Group #find should find single group by id  1`] = `
Object {
  "description": "Manage stuff",
  "id": 2,
  "label": "Manager",
  "name": "manager",
}
`;

exports[`Group #find should find single group by name 1`] = `
Object {
  "description": "Manage stuff",
  "id": 2,
  "label": "Manager",
  "name": "manager",
}
`;

exports[`Group #findMany should find many groups by id  1`] = `
Array [
  Object {
    "description": "",
    "id": 1,
    "label": "Admin",
    "name": "admin",
  },
  Object {
    "description": "Manage stuff",
    "id": 2,
    "label": "Manager",
    "name": "manager",
  },
]
`;

exports[`Group #findMany should find many groups by id  2`] = `
Array [
  Object {
    "description": "Manage stuff",
    "id": 2,
    "label": "Manager",
    "name": "manager",
  },
]
`;

exports[`Group #findMany should find many groups by name 1`] = `
Object {
  "description": "Manage stuff",
  "id": 2,
  "label": "Manager",
  "name": "manager",
}
`;

exports[`Group #getUserGroups should return list of user groups 1`] = `
Array [
  Object {
    "description": "",
    "id": 1,
    "label": "Admin",
    "name": "admin",
  },
  Object {
    "description": "Manage stuff",
    "id": 2,
    "label": "Manager",
    "name": "manager",
  },
]
`;

exports[`Group #isUserAttahcedTo should return false when user doesn't belong to a group 1`] = `false`;

exports[`Group #isUserAttahcedTo should return true when user belongs to a group 1`] = `true`;

exports[`Group #list should get list of groups 1`] = `
Array [
  Object {
    "description": "",
    "id": 1,
    "label": "Admin",
    "name": "admin",
  },
  Object {
    "description": "Manage stuff",
    "id": 2,
    "label": "Manager",
    "name": "manager",
  },
]
`;

exports[`Group #update should update a group by id 1`] = `
Object {
  "description": "",
  "id": 1,
  "label": "Admin",
  "name": "admin",
}
`;
